socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,492 @@
+            When the PID freq drops below this threshold, tell the hwp_pmx to cut power.
+        use_pid : bool
+            If True, will use the PID to spin down the HWP. If False, will
+            just tell the hwp_pmx agent to cut power.
+
+        """
+        clients = self._get_hwp_clients()
+
+        if not params.use_pid:
+            clients.pmx.set_off()
+            session.data = {
+                'pid_freq': None,
+                'pid_freq_thresh': params['pid_freq_thresh'],
+                'use_pid': params['use_pid']
+            }
+            return True, "Commanded HWP to stop spinning"
+
+        clients.pid.tune_stop()
+        clients.pmx.use_ext()
+        clients.pmx.set_on()
+        session.add_messsage("Tuning PID to stop")
+
+        session.data = {
+            'pid_freq_thresh': params.pid_freq_thresh,
+            'use_pid': params.use_pid
+        }
+        while True:
+            _, _, s = clients.pid.get_freq()
+            cur_freq = s.data['current_freq']
+            session.data['pid_freq'] = cur_freq
+            if cur_freq < params.pid_freq_thresh:
+                break
+
+        clients.pmx.set_off()
+        return True, "Commanded HWP to stop spinning"
+
+
+def make_parser(parser=None):
+    if parser is None:
+        parser = argparse.ArgumentParser()
+    pgroup = parser.add_argument_group('Agent Options')
+
+    pgroup.add_argument('--sleep-time', type=float, default=2.)
+
+    pgroup.add_argument('--hwp-lakeshore-id',
+                        help="Instance ID for lakeshore reading out HWP temp")
+    pgroup.add_argument('--hwp-temp-field',
+                        help='Field name of lakeshore channel reading out HWP temp')
+    pgroup.add_argument('--hwp-temp-thresh', type=float,
+                        help="Threshold for HWP temp.")
+
+    pgroup.add_argument('--hwp-encoder-id',
+                        help="Instance id for HWP encoder agent")
+    pgroup.add_argument('--hwp-pmx-id',
+                        help="Instance ID for HWP pmx agent")
+    pgroup.add_argument('--hwp-pid-id',
+                        help="Instance ID for HWP pid agent")
+    pgroup.add_argument('--ups-id', help="Instance ID for UPS agent")
+    pgroup.add_argument('--ups-minutes-remaining-thresh', type=float, default=30.,
+                        help="Threshold for UPS minutes remaining before a "
+                             "shutdown is triggered")
+
+    return parser
+
+
+def main(args=None):
+    parser = make_parser()
+    args = site_config.parse_args(agent_class='HWPSupervisor',
+                                  parser=parser,
+                                  args=args)
+
+    agent, runner = ocs_agent.init_site_agent(args)
+    hwp = HWPSupervisor(agent, args)
+
+    agent.register_process('monitor', hwp.monitor, hwp._stop_monitor,
+                           startup=True)
+    agent.register_task('spin_up', hwp.spin_up)
+    agent.register_task('spin_down', hwp.spin_down)
+
```

Comment:
Currently spin_up and spin_down are tasks, and contain potentially long-running, uninterruptible loops.  That's a problem.  There's also just a lot of duplication in those two functions.

I propose that instead of separate tasks, a process be used to manage the spin.  The process would normally be running, and would be smart about special speeds (such as stopping).  There would be separate, ~instantaneous task(s) available to change the target frequency / request stop.